Add support for IPv6 Virtual DNS (#462)

3nprob opened the pull request:
- #462 resurrected
- Fix merge conflicts
- Address outstanding review feedback
- Unify logic for generating firewall rules for ipv4/ipv6

Fixes https://github.com/QubesOS/qubes-issues/issues/10038

marmarek commented:
Now that the smaller PRs are in (the last one is still going through CI), this will need a rebase (dropping those already merged commits) to resolve conflicts.

Commit:
Add support for IPv6 Virtual DNS

- Merge '1cho1ce/add-ipv6-dnat-to-ns' into master
  - https://github.com/QubesOS/qubes-core-agent-linux/pull/462

- fix: properly assign primary/secondary DNS

- fix: check ipv4 dns presence by qdb /qubes-primary-dns instead of /qubes-ip

- qubes-setup-dnat-to-ns: unify ipv4/ipv6 firewall rule generation

Part of https://github.com/QubesOS/qubes-issues/issues/10038

Commit:
setup-ip: oneshot resolv.conf update

marmarek commented on qubes-rpc/post-install.d/10-qubes-core-agent-features.sh:
```diff
@@ -119,6 +119,8 @@ advertise_systemd_service() {
     done
 }
 
+qvm-features-request supported-feature.ipv6dns=1
```
It is referenced in the core-admin PR.

Commit:
chore: remove unused argument to install_firewall_rules

Commit:
fix nft_cmd construction, adding typing annotation

Commit:
revert conditional insertion of host-unreachable response

marmarek commented on network/qubes-setup-dnat-to-ns:
```diff
@@ -71,75 +72,89 @@ def get_dns_resolved():
+                    rules += [
+                        f"ip{ip46} daddr {vm_nameserver} tcp dport 53 reject with icmp{ip46} type host-unreachable",
+                        f"ip{ip46} daddr {vm_nameserver} udp dport 53 reject with icmp{ip46} type host-unreachable",
+                    ]
```
See the discussion on the original PR: https://github.com/QubesOS/qubes-core-agent-linux/pull/462
The issue is if you have only IPv4 or only IPv6 DNS - then you wouldn't have address of such kind. And the reject rule is to avoid long timeouts when trying non-existing DNS and immediately fallback to the other one.

If you have a way to test it, try with:
- only IPv4 DNS present
- only IPv6 DNS present
- both present

In all the cases, name resolution should keep working instantly. The tests I added in core-admin PR try to exercise those cases, but I'm not 100% if it will fail on slow fallback...

ArrayBolt3 commented on network/qubes-setup-dnat-to-ns:
```diff
@@ -74,74 +75,82 @@ def get_dns_resolved():
+            f"delete chain ip{ip46} qubes dnat-dns",
+        ]
+        rules = [
+            f"table ip{ip46} qubes {{",
```
I'm think the `{{` here is wrong.

```suggestion
            f"table ip{ip46} qubes {",
```

ArrayBolt3 replied:
Oh. Sorry, my mistake.

ArrayBolt3 commented:
Thanks! I think I can test this since my network supports both IPv4 and IPv6 and I can disable IPv6 for testing. I don't need push rights, I'll just open new PRs, but thank you :) And thanks for everything you and 1cho1ce did to get things this far!

ArrayBolt3 commented:
@3nprob From some discussion in the Qubes OS Matrix room, currently systemd-resolved is intentionally being bypassed as much as possible, because upstream doesn't recommend exposing it to the LAN, and so it's unclear whether it can safely handle potentially malicious clients. A different DNS resolver that is designed to work in this way might be an option, but then you have to run a DNS resolver in sys-net which is painful because sys-net is already severely resource-constrained by design.

IIUC, socat would allow us to simply DNAT the virtual IPv6 DNS address to sys-net's internal network adapter's IP address, then we can forward it to the real DNS server thereafter. It's crude, but as long as all socat does is pass bytes from point A to point B and then back again, it should work and not introduce any further attack surface. (Then again, I don't know what socat's attack surface looks like; I'd assume it's pretty close to zero, but maybe not?)

@marmarek Since it doesn't look like DNAT alone is going to work, what are your thoughts on adding socat to the solution?

marmarek commented:
Using socat enables one more trick - it doesn't care about protocol on the listening and sending side, so you can translate IPv4 to IPv6 and back. So, technically we could have DNS1 and DNS2 addresses in v4 and v6 flavors, and they would translate to upstream DNS1 and DNS2 regardless of which protocol they use. On one hand, it's tempting, since it would avoid needing every qube to try IPv6 first just to learn it isn't available at the time (when you're connected to IPv4-only network). But on the other hand, it feels like a lot could go wrong...

I did used something similar before, to be able to use IPv6-only network on qubes - simply socat listening on 10.139.1.1 and forwarding to whatever IPv6 DNS was there. But I don't have that script anymore, so I can't check what I did about listening address... I guess went with a dummy interface for it.

marmarek commented:
While adding this feature, it would be nice to consider already a case where there is only IPv6 traffic. Just because it's rather awkward to have need IPv4 just for (internal) DNS when everything else is on IPv6 already. On the other hand, there s still a very long way to IPv6-only internet, so maybe it doesn't really matter yet...

v3xro commented:
> While adding this feature, it would be nice to consider already a case where there is only IPv6 traffic. Just because it's rather awkward to have need IPv4 just for (internal) DNS when everything else is on IPv6 already. On the other hand, there s still a very long way to IPv6-only internet, so maybe it doesn't really matter yet...

Some UK mobile networks have now gone IPv6 only (personal experience) so tethering to a phone on the go no longer results in working networking anywhere except `sys-net`. So that's already a strong motivation to get this working (for me at least).
